### PR TITLE
Feature #2152 poly_xy

### LINKED ISSUE #2152: Enhance Gen-Vx-Mask by adding a new poly_xy masking type option.
## Describe the New Feature ##
This is a request from the UK MetOffice made during the METplus MetOffice telecon on 5/5/2022. @robdarvell ran gen_vx_mask on a rotated lat/lon grid and was surprised by the result. Using a polyline consisting of 4 points, he found that the edges of the resulting mask were curved rather than straight lines.

The existing ```-type poly``` option is performed in lat/lon space. The user provides a polyline consisting of lat/lon points. gen_vx_mask loops through all of the grid points, converts each from (x, y) to (lat, lon) space, and determines if that (lat, lon) point falls inside the lat/lon polyline.

The task here is to add a new option for ```-type poly_xy```. While the user still provides a polyline consisting of lat/lon points, those points are converted into grid (x, y) space. gen_vx_mask loops through all of the grid points and determines if each (x, y) grid point falls inside the x/y polyline.

The former produces straight lines on the earth while the latter produces straight lines on the grid. Both are reasonable approaches.

Also note that the MetOffice would like to add a ```coordinates``` variable attribute to gridded NetCDF output written by MET. That change would apply to masks written by Gen-Vx-Mask as well as the many other MET tools that write gridded NetCDF output.

### Acceptance Testing ###
- Add new ```-type poly_xy``` option and add a unit test to demonstrate it.
- Confirm that the gridded NetCDF output files written by MET include the ```coordinates = "lat lon";``` variable attribute.

### Time Estimate ###
4 hours.

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
None needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
MetOffice support key (2799991)

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: @JohnHalleyGotway 
- [x] Select **scientist(s)** or **no scientist** required: @robdarvell 

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- [x] Coordinate with @georgemccabe to determine if an update to the Gen-Vx-Mask wrapper is needed.

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [x] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
Note that this PR does 3 main things:
1. Adds support for gen_vx_mask -type poly_xy option.
2. Adds the coordinates NetCDF variable attribute under the right set of circumstances.
3. Improves the log message written by the check_prob_thresh() function.

OLD:
```
ERROR  : 
ERROR  : check_prob_thresh() -> When verifying a probability field, you must select at least 3 thresholds beginning with 0.0 and ending with 1.0.
ERROR  : 
```

NEW:
```
ERROR  : 
ERROR  : check_prob_thresh() -> When verifying a probability field, you must select at least 3 thresholds beginning with 0.0 and ending with 1.0 (current setting: >=0,>=0.5,>=1.1).
ERROR  : Consider using the ==p shorthand notation for bins of equal width.
ERROR  : 
```

## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>
Adds support for a new command line option in Gen-Vx-Mask: ```-type poly_xy```

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>
For NetCDF output files from MET, if the variable contains both the "lat" and "lon" dimensions and is also not named "lat" or "lon", a new coordinates variable attribute is written:
```
	float CONUS(lat, lon) ;
		CONUS:coordinates = "lat lon" ;
```
This is added for the MetOffice NetCDF visualization software they use.

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Manually tested and provided output images in this [issue comment](https://github.com/dtcenter/MET/issues/2152#issuecomment-1118872179).

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
@georgemccabe: Please review the Gen-Vx-Mask code changes, new units test, documentation updates, and updated check_prob_thresh() error messages. Does the Gen-Vx-Mask wrapper need to be updated to support the poly_xy option?
@hsoh-u: Please review the vx_util logic for the addition of the coordinates variable attribute.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [x] Do these changes include sufficient testing updates? **[Yes]**
Added 2 new calls to gen_vx_mask in unit_gen_vx_mask.xml.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
- 2 new mask files will be added.
- Many, many NetCDF output files will be modified with the addition of the coordinates variable attribute.

- [x] Please complete this pull request review by **[Mon 5/9/22]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Review by georgemccabe — APPROVED** [review]
I reviewed the changes for gen_vx_mask and they look good. The diff for the documentation is hard to follow because it highlights many lines that are not actually changed due to the list numbering, but it looks like the Polyline XY description was added and the Polyline description was slightly modified. The code changes to add support for the poly_xy type are clear and well documented.

I didn't see the "compliment" variable set in the function, so I assume that it is a class variable. Some coding standards require class variables to start with an underscore, so that is something to consider for another existing GitHub issue.

The GenVxMask wrapper supports a single variable to set all of the command line arguments for the tool. The type is set in this way, so no changes to the wrapper to support the new setting are needed.

The updates to the error logs are more clear and helpful.

I approve.

File: met/docs/Users_Guide/masking.rst [hunk 1]
```diff
@@ -46,7 +46,7 @@ Required arguments for gen_vx_mask
 
 2. The **mask_file** argument defines the masking information, see below.
 
-• For "poly", "box", "circle", and "track" masking, specify an ASCII Lat/Lon file.
+• For "poly", "poly_xy", """box", "circle", and "track" masking, specify an ASCII Lat/Lon file.
```
Comment (hsoh-u) [inline met/docs/Users_Guide/masking.rst]: three double quotation marks: """box"

**Review by hsoh-u — APPROVED** [review]
I approve . Only write_netcdf.cc calls add_var with the vector of dimensions. The variable names and the number of dimensions match at write_netcdf.cc and add_var().

File: met/src/libcode/vx_nc_util/nc_utils.cc [hunk 2]
```diff
@@ -3216,6 +3214,26 @@ NcVar add_var(NcFile *nc, const string &var_name, const NcType ncType,
...
+      else if (itDim->getName() == "lon") has_lon_dim = true;
+      if (itDim != ncDims.begin()) cs << " ";
+      cs << itDim->getName();
+   }
+
+   // Add the coordinates variable attribute for variables
+   // with both lat and lon dimensions
+   if (has_lat_dim && var_name != "lat" &&
+       has_lon_dim && var_name != "lon") {
+      add_att(&var, "coordinates", cs.c_str());
```
Comment (hsoh-u) [inline met/src/libcode/vx_nc_util/nc_utils.cc:3234]: Comment: The coordinate attribute is added only if the lat/lon are 2D.

File: met/docs/Users_Guide/masking.rst (hunk 1 above)
Comment (JohnHalleyGotway) [reply met/docs/Users_Guide/masking.rst]: Thanks @hsoh-u, I fixed that one.

File: met/src/libcode/vx_nc_util/nc_utils.cc (hunk 2 above)
Comment (JohnHalleyGotway) [reply met/src/libcode/vx_nc_util/nc_utils.cc:3234]: Yes, that is by design. When I searched for example of the use of the "coordinates" attribute, I only saw it used for variables that include both latitude and longitude dimensions.

**Comment by JohnHalleyGotway** [comment]
As expected, the GHA testing workflow flagged diffs. I retrieved and inspected the comp_dir.log file from [logs](https://github.com/dtcenter/MET/suites/6395484154/artifacts/232674916). 

As expected, there are 2 new output files:
```
ERROR: folder /data/output/met_test_truth missing 2 files
    gen_vx_mask/POLY_LATLON_RECTANGLE_mask.nc 
    gen_vx_mask/POLY_XY_RECTANGLE_mask.nc 
```

And the header is modified in 223 NetCDF output files:
```
egrep -i "file1:|ERROR:" comp_dir.log  | egrep -B 1 ERROR | grep file1 | wc -l
     223
```
The changes are all the addition of the coordinates attribute:
```
TMP_Z2_ENS_MEAN:coordinates = "lat lon" ;
```
And these are added to 2265 NetCDF variables!
```
grep "coordinates" comp_dir.log  | wc -l
    2265
```
